apis/duck/v1alpha1/addressable_types.go

```diff
@@ -26,15 +26,17 @@ import (
 
 // Addressable provides a generic mechanism for a custom resource
 // definition to indicate a destination for message delivery.
-// (Currently, only hostname is supported, and HTTP is implied. In the
-// future, additional schemes may be supported, and path components
+// (Currently, only hostname, port and path are supported, and HTTP is implied. In the
+// future, additional schemes may be supported
 // ala UI may also be supported.)
 
 // Addressable is the schema for the destination information. This is
 // typically stored in the object's `status`, as this information may
 // be generated by the controller.
 type Addressable struct {
 	Hostname string `json:"hostname,omitempty"`
+	Port     string `json:"port,omitempty"`
+	Path     string `json:"path,omitempty"`
```

Comment:
My preference is to have the new fields be on a struct that is a sibling of the existing Hostname. 

```golang
type Addressable struct {
	Hostname string `json:"hostname,omitempty"`
	Address *Address `json:"address,omitempty"` // terrible name, replace with something better.
}

type Address struct {
	Hostname string `json:"hostname,omitempty"`
	Port     string `json:"port,omitempty"`
	Path     string `json:"path,omitempty"`
}
```

Or just make it a `*url.URL`.

My reasoning is that this is a breaking change. Any old code that is resolving an Addressable that uses path or port will create the wrong URI. 

For example:

```yaml
# Current PR Proposal
...
status:
  addressable:
    hostname: foo
    path: bar
---
# Sibling
...
status:
  addressable:
    address:
      hostname: foo
      path: bar
```

These should produce an address of `http://foo/bar`.

If we add the fields as proposed, then the old code will create an incorrect URI, `http://foo/`. It won't realize that the path field is important. This looks valid and we will only see that it fails when the request is actually made (and we _really_ hope the remote server doesn't respond successfully to it). 

If we use the sibling `Address` instead, then the old code will notice that `status.addressable.hostname` is blank, so it is not yet ready (well, we hope that the code doesn't just try `http:///`). This should manifest as a visible error either on the status or a Kubernetes event associated with the resource that needs the resolution done.

---

In either case, this change is backwards incompatible. In the current proposal, we will only notice the error when the request is made (and hope it doesn't do anything bad). In the sibling proposal, we should see the error more clearly in some control plane messages. In the worst case, it is identical to the current proposal.

Reply:
We will have to do it the way you proposed while we migrate all eventsources and then finally get rid of istio from eventing.